### PR TITLE
dom api implementation

### PR DESCRIPTION
**Pull request by ptol**
resolves https://github.com/paldepind/snabbdom/issues/47

### CODE REVIEW

**Comment by ptol** [comment]
The idea is anyone can implement this interface
```
module.exports = {
  createElement: createElement,
  createElementNS: createElementNS,
  createTextNode: createTextNode,
  appendChild: appendChild,
  removeChild: removeChild,
  insertBefore: insertBefore,
  parentNode: parentNode,
  nextSibling: nextSibling,
  tagName: tagName,
  setTextContent: setTextContent
};
```
and use snabbdom with any type of dom.

**Comment by rayd** [comment]
I like this. Similar to the changes I made to add <iframe> support (see #72), but more flexible. I think I could re-implement what's in #72 without having to change the core for people who don't care about iframes.

**Comment by TylorS** [comment]
I'll be that guy who does a :+1:

File: snabbdom.js [hunk 1]
```diff
@@ -232,7 +237,7 @@ function init(modules) {
     var insertedVnodeQueue = [];
     for (i = 0; i < cbs.pre.length; ++i) cbs.pre[i]();
 
-    if (oldVnode.nodeType === Node.ELEMENT_NODE) {
+    if (isUndef(oldVnode.sel)) {
```
Comment (paldepind) [inline snabbdom.js]: Nice workaround here!

**Comment by paldepind** [comment]
This is nice. What are you planning to use this for @ptol?

I'm a bit hesitant to merge this as I'm not sure if and how people will use it?

@rayd That sounds like a really good idea. Iframe support is nice but also rarely needed.

**Comment by ericgj** [comment]
:+1: from me.  Was just looking at the feasibility of rendering vnodes to something like [blessed](https://github.com/chjj/blessed) (ncurses), this would help I think.

**Comment by ptol** [comment]
@paldepind There are dozen of js libraries with their own dom (all game engines and canvas/webgl libraries) so if someone need a virtual dom for such library the snabbdom would be a possible solution.

I have been using it with pixi.js
```
var nodeTypes = {};
nodeTypes.text = () => new PIXI.Text("", {font: "14px Verdana", fill: "white"});
nodeTypes.group = () => new PIXI.Container();
nodeTypes.sprite = () => new PIXI.Sprite(px.Texture.EMPTY);
...

function createElement(tagName){
  var node = nodeTypes[tagName]();
  if(node){
    node.tagName = tagName;
  }
  return node;
}

function insertBefore(parentNode, newNode, referenceNode){
    if(referenceNode){
      var i = parentNode.getChildIndex(referenceNode);
      parentNode.addChildAt(newNode,i);
    }else{
      parentNode.addChild(newNode);
    }
}

function removeChild(node, child){
  node.removeChild(child);
}

function appendChild(node, child){
  node.addChild(child);
}

function parentNode(node){
  return node.parent;
}

function nextSibling(node){
  var parent = node.parent;
  var i = parent.getChildIndex(node);
  var sublingIndex = i + 1;
  if(parent.children.length <= sublingIndex) return null;
  return parent.getChildAt(sublingIndex);
}

function tagName(node){
  return node.tagName;
}

module.exports = {
  createElement: createElement,
  appendChild: appendChild,
  removeChild: removeChild,
  insertBefore: insertBefore,
  parentNode: parentNode,
  nextSibling: nextSibling,
  tagName: tagName
};
```

**Comment by ptol** [comment]
@paldepind Or do you have something different in mind for how this https://github.com/paldepind/snabbdom/issues/47 could be solved?

**Comment by paldepind** [comment]
@ptol Nope. This definitely seems to be the way to solve #47. The above is exactly what I wanted to see. A proof of concept that this abstracted interface can actually be usefully implemented by anything besides the DOM. Using it to control the pixi.js scene graph is certainly useful.

File: snabbdom.js [hunk 2]
```diff
@@ -30,14 +31,18 @@ function createKeyToOldIdx(children, beginIdx, endIdx) {
...
+      var parent = api.parentNode(childElm);
+      api.removeChild(parent, childElm);
+    }
   };
 }
 
 var hooks = ['create', 'update', 'remove', 'destroy', 'pre', 'post'];
 
-function init(modules) {
+function init(modules, domApi) {
```
Comment (paldepind) [inline snabbdom.js]: This is nitpicking but I think the variable names are slightly confusing. We call the actual DOM api `api` and the user supplied API `domApi`. But the user supplied interface implementation might have nothing to do with the DOM while the DOM interface certainly does.

What about renaming `api` to `domApi` and `domApi` to `userApi`. Alternatively rename `api` to `domImpl` and `domApi` to `apiImpl`. The former is probably best though.

Comment (ptol) [inline snabbdom.js]: I don't mind. You can change any names you think would be more suitable. For example, I changed parentElement to parentNode but maybe it should stay parentElement.
```
function parentNode(node){
  return node.parentElement;
}
```

Comment (paldepind) [inline snabbdom.js]: I think `parentNode` is good.

**Comment by paldepind** [comment]
@ptol Thank you a lot for this PR! As you can see I've merged it.

Unless I'm mistaken then the code in this PR would force all the `patch` functions returned by `init` to use the same DOM API. I've made a commit that should keep the different APIs separate. So that one, for instance, can use Snabbdom to patch both the DOM and pixi.js on the same page with different `patch` instances. Will you try it ou @ptol? After that I'll do a release.

Thanks again :smile:

**Comment by ptol** [comment]
>Unless I'm mistaken then the code in this PR would force all the patch functions returned by init to use the same DOM API. 
I don't think so,  `domApi` is a local variable 
```
var domApi = require('./htmldomapi.js');
```
so
```
var patch1 = snabbdom.init([], require('./pixiapi')); //pixi
var patch2 = snabbdom.init([]); //html
```
shouldn't affect each other.

> I've made a commit that should keep the different APIs separate

Do you mean this commit https://github.com/paldepind/snabbdom/commit/fb80c1fd80235e93803d51a7ab274d659fd1a317 ?
I don't see any related changes.

**Comment by ericgj** [comment]
No strong feelings about it, but what about a new `patch` constructor instead of overloading `init` ?  Also to me it feels like the api should be the first parameter, and the modules the second. 

```js
var patch1 = snabbdom.initWithAPI( api, modules );
var patch2 = snabbdom.init( modules );   // === snabbdom.initWithAPI( htmldomapi, modules );
```

**Comment by paldepind** [comment]
> I don't think so,  `domApi` is a local variable 

But `api` is local to the entire module t

> Do you mean this commit https://github.com/paldepind/snabbdom/commit/fb80c1fd80235e93803d51a7ab274d659fd1a317 ?
I don't see any related changes.

I've changed it so that `api` is local to the function returned by `patch` and move a few function that use `api` so that they are inside the created closure.

@ericgj

> No strong feelings about it, but what about a new `patch` constructor instead of overloading `init` ?  Also to me it feels like the api should be the first parameter, and the modules the second. 

A new `initWithApi` function wouldn't bother me—but I'm indifferent. I don't mind the overloading here but I can see the appeal in avoiding it. Why do you feel the API argument should be first?

**Comment by ptol** [comment]
Oh, I see. api should definitely be in function scope. I used last version with different dom on one page and everything work fine.

**Comment by paldepind** [comment]
Released and published to npm.

@ptol If you feel like it I think publishing your pixi.js implementation of the DOM interface would be really great! We can create a folder in the repo for API implementations or you can create a repository of your own and I will link to it from the Snabbdom README.

**Comment by ptol** [comment]
@paldepind Yes, I have plans to create a repository but I need to fix some problems before.
In canvas/webgl libraries there is a method `destroy`. This method need to be called when you don't need the object anymore. 

```
var vnode1 = h("group", {}, [h("text", {props: {text: "foo"}} )]);
var vnode1 = h("group", {}, []);
```
For example, this text object should be destroyed. What would be the best place to call this method? 
Can I do this?

```
function removeChild(node, child){
  node.removeChild(child);
  child.destory();
}
```
Or even it's removed this child can still be used somewhere?

**Comment by ptol** [comment]
Here the repository https://github.com/ptol/virtual-pixi 
It seems there isn't any problem with 
```
function removeChild(node, child){
  node.removeChild(child);
  child.destory();
}
```

**Comment by Swizz** [comment]
Hi,

This is a great valuable addition which require to be shown in the doc.
I use snabbdom for a while, however this is our google friend that driven me on that issue when I was looking for virtualdom tool for NativeScript.

Cheers,

**Pull request merged**